### DISCUSSION TITLE
Generic where clause

### DISCUSSION THREAD

**Discussion by josham** [sqlalchemy/sqlalchemy#11988] [Usage Questions]
I am wondering if it's possible to take a where clause on an ORM mapped class but change its alias later (since it isn't know ahead of time).
In the example, I'm trying to change the where clause from using User and Address to u and a respectively. Or, another way to put it, is if it's possible to have an expression on columns not associated with a table, e.g. `w = col("name") == "foo"` and then apply them to whatever column matches the name in the select e.g. `select(User.name).where(w)`

I've also asked this question in gitter -- not sure the right place so posting here as well.

Thanks for any help.

```python
from sqlalchemy import String, select, or_, ForeignKey
from sqlalchemy.orm import DeclarativeBase, aliased, Mapped, mapped_column, relationship

class Base(DeclarativeBase):
    pass

class User(Base):
    __tablename__ = "user_account"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30))
    addresses: Mapped[list["Address"]] = relationship(
        back_populates="user", cascade="all, delete-orphan"
    )

class Address(Base):
    __tablename__ = "address"
    id: Mapped[int] = mapped_column(primary_key=True)
    email_address: Mapped[str]
    user_id: Mapped[int] = mapped_column(ForeignKey("user_account.id"))
    user: Mapped["User"] = relationship(back_populates="addresses")


def filter_by(where):
    u = aliased(User, name="u")
    a = aliased(Address, name="a")
    s = select(u.name, a.email_address).join(a.user.of_type(u))
    return s.where(where)


where = or_(User.name == "foo", Address.email_address == "foo@bar.com")
s = filter_by(where)
```

**Comment by zzzeek** [comment]
we have something that's sort of close to this which is [with_loader_criteria](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria), however that's tailored towards limiting a query for certain full entities of a specific mapper.

in this example I can take you straight behind the curtain and give you the API SQLAlchemy uses internally for this, which truth be told you might be the first person in all of SQLAlchemy's 18 year history to ask for, but then again our API is not like it used to be.   You can "adapt" any SQL fragment to convert FROM objects like this:

```py
def filter_by(where):
    u = aliased(User, name="u")
    a = aliased(Address, name="a")
    s = select(u.name, a.email_address).join(a.user.of_type(u))
    from sqlalchemy import inspect
    where = inspect(u)._adapt_element(where)
    where = inspect(a)._adapt_element(where)
    return s.where(where)

```


so, there's that, however, as you can see the API here is not really public.   There are public-ish ways to get at the above but the "_adapter" that's local to the `aliased()` object has the best configuration set up, and that's not "public" because as mentioned nobody ever asks for this, it's a bit esoteric and while I tried to get people [interested in the idea of clause adaptation sixteen years ago](https://techspot.zzzeek.org/2008/01/23/expression-transformations/), wondering if I had stumbled onto something already known or generally useful, nobody expressed any interest in this core feature of SQLAlchemy's ORM. 

That might be because there's usually simpler ways to do this in user-land.   In this example, a more portable way to do this is to just defer the creation of your `or_()` using this approach:

```py
def filter_by(where):
    u = aliased(User, name="u")
    a = aliased(Address, name="a")
    s = select(u.name, a.email_address).join(a.user.of_type(u))
    return s.where(where(User=u, Address=a))


where = lambda User, Address: or_(User.name == "foo", Address.email_address == "foo@bar.com")
s = filter_by(where)
print(s)
```

I would prefer to do it that way for an end-user program, simple and explicit.

**Reply by josham** [reply]
Thanks for the response and happy to hear I have an original question :) Seems like `_adapt_element` is close to what I want but  on first pass it doesn't seem to give me enough control -- in my use case I have the same table aliased multiple times. 

Let me try to do a better job of explaining my problem. I'm looking to provide a user API where they can supply the filter criteria without having to know about where the columns originate. In some cases they are columns in a table but in others, the `column` is the result of joining/filtering on another table. 

One solution (though this may not be possible) would be to define columns without an associated table and then later bind them to a particular column in the expression. In my use case, I would inspect the `where` to determine the columns they are filtering on, and build up the query specifically to handle that subset of columns.

I may be able to accomplish a less general version of this using `column_property`, looking at that now as well.

```python
USER_ID = Column("user_id", INT)
USER_NAME = Column("user_name", String(30))
USER_EMAIL = Column("user_email", String(30))

def filter_by(where):
    u = aliased(User, name="u")
    a = aliased(Address, name="a")
    s = select(u.name.label("user_name"), a.email_address.label("user_email")).join(a.user.of_type(u))
    return s.where(where)

where = or_(USER_NAME == "foo", USER_EMAIL == "foo@bar.com")
s = filter_by(where)
```

**Reply by CaselIT** [reply]
> in my use case I have the same table aliased multiple times.

if you have multiple alias I would expect the user of your api to know that there are multiple alias. Otherwise they couldn't target a particular alias.
Given this a function is likely the easiest api. Pass it with a dict alias_name->alias_object and the let user create the expression they like with them
